plugins/automelt.cpp

```diff
@@ -1,294 +1,705 @@
-#include "uicommon.h"
 
-#include "modules/Gui.h"
+#include "Debug.h"
+#include "LuaTools.h"
+#include "PluginManager.h"
+#include "TileTypes.h"
+
+#include "modules/Buildings.h"
+#include "modules/Maps.h"
+#include "modules/Items.h"
+#include "modules/World.h"
+#include "modules/Designations.h"
+#include "modules/Persistence.h"
+#include "modules/Units.h"
+
+// #include "uicommon.h"
 
 #include "df/world.h"
+#include "df/building.h"
 #include "df/world_raws.h"
 #include "df/building_def.h"
 #include "df/viewscreen_dwarfmodest.h"
 #include "df/building_stockpilest.h"
-#include "modules/Buildings.h"
-#include "modules/Items.h"
 #include "df/plotinfost.h"
-#include "modules/Maps.h"
-#include "modules/World.h"
 #include "df/item_quality.h"
 
+#include <map>
+#include <unordered_map>
+
 using df::building_stockpilest;
+using std::map;
+using std::multimap;
+using std::pair;
+using std::string;
+using std::unordered_map;
+using std::vector;
+
+using namespace DFHack;
+using namespace df::enums;
 
 DFHACK_PLUGIN("automelt");
-#define PLUGIN_VERSION 0.3
 REQUIRE_GLOBAL(gps);
 REQUIRE_GLOBAL(world);
 REQUIRE_GLOBAL(cursor);
 REQUIRE_GLOBAL(plotinfo);
 
-static const string PERSISTENCE_KEY = "automelt/stockpiles";
+namespace DFHack
+{
+    DBG_DECLARE(automelt, status, DebugCategory::LINFO);
+    DBG_DECLARE(automelt, cycle, DebugCategory::LINFO);
+    DBG_DECLARE(automelt, perf, DebugCategory::LTRACE);
+}
+
+static const string CONFIG_KEY = string(plugin_name) + "/config";
+static const string STOCKPILE_CONFIG_KEY_PREFIX = string(plugin_name) + "/stockpile/";
+static PersistentDataItem config;
+
+static vector<PersistentDataItem> watched_stockpiles;
+static unordered_map<int, size_t> watched_stockpiles_indices;
+
 
-static int mark_item(df::item *item, df::item_flags bad_flags, int32_t stockpile_id)
+enum ConfigValues
 {
-    if (item->flags.whole & bad_flags.whole)
-        return 0;
+    CONFIG_IS_ENABLED = 0,
 
-    if (item->isAssignedToThisStockpile(stockpile_id)) {
-        size_t marked_count = 0;
-        std::vector<df::item*> contents;
-        Items::getContainedItems(item, &contents);
-        for (auto child = contents.begin(); child != contents.end(); child++)
-        {
-            marked_count += mark_item(*child, bad_flags, stockpile_id);
+};
+
+enum StockpileConfigValues
+{
+    STOCKPILE_CONFIG_ID = 0,
+    STOCKPILE_CONFIG_MONITORED = 1,
```

Comment:
I believe thresholds for melting objects are better handled in the manager orders, not here. I'm not sure what an extra `monitored` flag would do for this plugin.

Reply:
reading the gui code.. I see that this can give you a tally of the number of items in monitored stockpiles. Could you explain the cases where this is handy for the player?

Reply:
So the plugin will only produce item counts if the pile is `monitored`, and will only melt if both `monitored` and `melt` are enabled.
The actual idea was a bit fuzzier, but would allow you to set a similar flag/value to autochop, in that it will only bother marking items in a `monitored` and `melt` enabled pile if the number of items to be melted is above some threshold, configurable via the UI. This would mean that if I'd had a work order set up to say `melt metal item` with the condition of `if 10 or more meltable items available` I wouldn't get order cancellation spam if there ended up being about 11 or so items marked. E.g, the config threshold value is "only mark for melting if there are 10 or more items to be marked" and the matching work order expects 10 items, the plugin would mark items in batches of 10. 

Further extension would mean you could set a config to leave a minimum number of items in enabled/marked stockpiles. Since the stockpile filters are a bit fiddly, you could end up with a situation where you want to keep at least x of some item around, but melt the excess. Instead of having to carefully set the stockpile size to respect that desired item count (and use pile links etc.) you can just set that in the pile's automelt config. It'd monitor the pile, and mark for melting above a certain count, then stop.

I have no idea if any of the above would actually be useful to anyone, even to me TBH. Can disable the extra flag for now, basically make the `STOCKPILE_CONFIG_MONITORED` flag pull double-duty as the melt-enable flag. Monitored piles get their items counted, all others are skipped in the initial cycle phase, so it cuts down on some exec time.

I also don't know if having the live count/overview of marked/monitored items in enabled piles is useful. I get the count numbers from the cycle pass anyway, and it produces that as output if you run `automelt designate` so I passed it to the GUI because it was easy.

Reply:
ok, I understand. I support having only one flag, if only because the use case is not well defined. Sending the count of melt-marked to automelt is absolutely useful. players can use that count to decide whether to build more smelters. I'd actually send the count of *all* melt-marked items, not just the ones in the stockpiles. I'm not sure what decision the stockpile-only count would be useful to make..

Reply:
Well, the way it's rigged now I only loop over items in the monitored stockpiles, so I don't have to go over every item present in the world. I could add a toggle-able option to scan all items everywhere, and return a count of melt-marked globally as well. Unsure if there's an already filtered list of items marked for melting, I vaguely remember seeing something like that though.

Latest ver. has only one flag to manage, `monitored`. Any stockpile with that enabled gets its items automatically marked and processed.

Reply:
It does look like there is an items vector `ANY_MELT_DESIGNATED`. that would make it cheap enough not to have to make it a special flag